Extension/src/githubAPI.ts

```diff
@@ -0,0 +1,152 @@
+function isBuild(input: any): input is Build {
+    const ok: boolean = input && input.name && typeof(input.name) === "string" &&
+        isArrayOfAssets(input.assets) && input.assets.length >= 4;
+    return ok;
+}
+
+function isArrayOfAssets(input: any): input is Asset[] {
+    const ok: boolean = input instanceof Array && input.every(item => isAsset(item));
+    return ok;
+}
+
+function isArrayOfBuilds(input: any): input is Build[] {
+    let ok: boolean =  input && input instanceof Array && input.length !== 0;
+    // Only check the five most recent builds for validity -- no need to check all of them
+    for (let i: number = 0; i < 5 && i < input.length; i++) {
+        if (!ok) {
+            return false;
+        }
+        ok = ok && isBuild(input[i]);
+    }
+    return ok;
+}
+
+function vsixNameForPlatform(info: PlatformInformation): string {
+    const vsixName: string = function(platformInfo): string {
+        switch (platformInfo.platform) {
+            case 'win32': return 'cpptools-win32.vsix';
+            case 'darwin': return 'cpptools-osx.vsix';
+            default: {
+                switch (platformInfo.architecture) {
+                    case 'x86': return 'cpptools-linux32.vsix';
+                    case 'x86_64': return 'cpptools-linux.vsix';
+                }
+            }
+        }
+    }(info);
+    if (!vsixName) {
+        const distro: string = info.distribution ? (':' + info.distribution.name + '-' + info.distribution.version) : '';
+        throw (new Error('Failed to match VSIX name for: ' + info.platform + distro + ':' + info.architecture));
+    }
+    return vsixName;
+}
+
+async function downloadUrlForPlatform(build: Build): Promise<string> {
+    // Get the VSIX name to search for in build
+    const info: PlatformInformation = await PlatformInformation.GetPlatformInformation();
+    const vsixName: string = vsixNameForPlatform(info);
+
+    // Get the URL to download the VSIX, using vsixName as a key
+    const downloadUrl: string = build.assets.find((asset) => {
+        return asset.name === vsixName;
+    }).browser_download_url;
+
+    if (!downloadUrl) {
+        return Promise.reject(new Error('Failed to find VSIX: ' + vsixName + ' in build: ' + build.name));
+    }
+    return downloadUrl;
+}
+
+export async function getTargetBuildURL(updateChannel: string): Promise<string> {
+    return getReleaseJson()
+        .then(builds => getTargetBuild(builds, updateChannel))
+        .then(build => { return build ? downloadUrlForPlatform(build) : Promise.resolve(undefined); } );
+}
+
+// Determines whether there exists a build that should be installed; returns the build if there is
+function getTargetBuild(builds: Build[], updateChannel: string): Build | undefined {
+    // Get predicates to determine the build to install, if any
+    let needsUpdate: (v1: PackageVersion, v2: PackageVersion) => boolean;
```

Comment:
Self-documentation would improve if the parameters were renamed: `v1` -> `installed` and `v2` -> `latest` (or similar)